spmd/tensor/placement_types.py

```diff
@@ -38,4 +39,46 @@ class Replicate(Placement):
 @dataclass
 class _Partial(Placement):
     # partial placement with reduce op
-    reduce_op: c10d.ReduceOp
+    reduce_op: c10d.ReduceOp = c10d.ReduceOp.SUM
+
+
+# used internally to propagate the placements
+@dataclass
+class PlacementSpec(object):
+    __slots__ = ["ndim", "mesh", "placements"]
+    ndim: int
+    mesh: DeviceMesh
+    placements: Sequence[Placement]
+
+    @property
+    def dim_map(self) -> List[int]:
+        # dims mapping of dist tensor sharding
+        # return size of tensor ndim, -1 represent replicate
+        # and int >=0 represent shard on that device mesh dim
+        r = [-1] * self.ndim
+        for i, placement in enumerate(self.placements):
+            if placement.is_shard():
+                shard_dim = cast(Shard, placement).dim
+                r[shard_dim] = i
+        return r
+
+    @classmethod
+    def from_dim_map(
```

Comment:
It would be nice that we can add more documentations on how to use this API and `dim_map`. Or if you want to merge this one first, we can add a TODO here for a detailed document which covers different scenarios on how users can use this API to define specs for examples like [Replicate(), Shard(0), Shard(1)], or [Shard(0), Shard(0)].

Reply:
added :)